pkg/placement/streams.go

```diff
@@ -0,0 +1,151 @@
+package placement
+
+import (
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"google.golang.org/grpc/metadata"
+
+	placementv1pb "github.com/dapr/dapr/pkg/proto/placement/v1"
+)
+
+type placementGRPCStream placementv1pb.Placement_ReportDaprStatusServer //nolint:nosnakecase
+
+type daprdStream struct {
+	id            uint32
+	hostName      string
+	hostID        string
+	hostNamespace string
+	needsVNodes   bool
+	stream        placementGRPCStream
+}
+
+func newDaprdStream(host *placementv1pb.Host, stream placementGRPCStream) *daprdStream {
+	return &daprdStream{
+		hostID:        host.GetId(),
+		hostName:      host.GetName(),
+		hostNamespace: host.GetNamespace(),
+		stream:        stream,
+		needsVNodes:   hostNeedsVNodes(stream),
+	}
+}
+
+// streamConnPool has the stream connections established between the placement gRPC server
+// and the Dapr runtime grouped by namespace, with an assigned id for faster lookup/deletion
+// The id is a simple auto-incrementing number, for efficiency.
+type streamConnPool struct {
+	lock sync.RWMutex // locks the streams map itself
+	// mutexMap *concurrency.MutexMap // manages locks on elements in the streams map
+
+	// Example representation of streams
+	//	{
+	//	  "ns1": {
+	//	  	1: stream1,
+	//	 	2: stream2,
+	//			3: stream3,
+	//		},
+	//
+	//	"ns2": {
+	//	  	4: stream4,
+	//	 	5: stream5,
+	//		},
+	//	}
+	streams map[string]map[uint32]*daprdStream
+
+	// streamIndexCnt assigns an index to streams in the streamConnPool.
+	// Its reset to zero every time a placement service loses leadership (thus clears all streams).
+	streamIndexCnt atomic.Uint32
```

Comment:
nit. Technically, this is not a count as streams are removes from the set but this number still goes up.
```suggestion
	// streamIndex assigns an index to streams in the streamConnPool.
	// Its reset to zero every time a placement service loses leadership (thus clears all streams).
	streamIndex atomic.Uint32
```